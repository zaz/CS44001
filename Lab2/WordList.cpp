#include "WordList.hpp"
#include <iostream>
#include <set>
#include <string>

std::set<std::string> WordList::uniqueWords() {
  return std::set<std::string>(wordCounts_.begin(), wordCounts_.end());
}

void WordList::addWord(const string& word) {
  wordCounts_.insert(word);
}

void WordList::print() {
  for (auto& word : uniqueWords())
    std::cout << wordCounts_.count(word) << ' ' << word << '\n';
}

unsigned int WordList::count(const std::string& word) {
  return wordCounts_.count(word);
}

bool equal(const WordList& w1, const WordList& w2) {
  return w1.wordCounts_ == w2.wordCounts_;
}
