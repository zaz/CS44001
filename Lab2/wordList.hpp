#ifndef WORDLIST_H_
#define WORDLIST_H_

#include <string>

using string = std::string;

class WordOccurrence {
public:
    WordOccurrence(const string& word="", int num=0);
    bool matchWord(const string &); // returns true if word matches stored
    void increment(); // increments number of occurrences
    string getWord() const;
    int getNum() const;

private:
    string word_;
    int num_;
};

class WordList{
public:
    // add copy constructor, destructor, overloaded assignment

    // implement comparison as friend
    friend bool equal(const WordList&, const WordList&);

    void addWord(const string &);
    void print();
private:
    WordOccurrence *wordArray_; // a dynamically allocated array of WordOccurrences
                                // may or may not be sorted
    int size_;
};


#endif // WORDLIST_H_
