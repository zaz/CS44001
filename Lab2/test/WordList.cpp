// testing the implementation of class WordList
// Mikhail Nesterenko
// 9/04/2015

#include <iostream>
#include <string>
#include "../WordList.hpp" // class definitions

using std::cout; using std::endl;

void testfunc(WordList); // function to test pass-by-value for WordList class

int main(){

   WordList w;

   // testing regular member functions
   w.addWord("one");
   w.addWord("one"); // adding duplicate
   w.addWord("two");
   w.addWord("three");

   cout << "word list is:\n";
   w.print();


   WordList w2, w3;

   w3=w2=w; // testing stacked overloaded assignment
   w3=w3; // testing protection against self-assingment

   if(equal(w2,w3))
      cout << "w2 is equal to w3" << endl;
   else
      cout << "w2 is not equal to w3" << endl;

   testfunc(w); // testing copy constructor
   w.print();   // if destructor is implemented correctly
                // this should print properly after testfunc completes
}


// tests pass-by-value for object of class varArray
void testfunc(WordList myList){ // copy constructor is invoked on "myList"
   cout << "word list in testfunc: " << endl;
   myList.print();
} // destructor is invoked when "myList" goes out of scope

