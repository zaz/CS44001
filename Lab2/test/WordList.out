word list is:
2 one
1 three
1 two
w2 is equal to w3
word list in testfunc: 
2 one
1 three
1 two
2 one
1 three
1 two
